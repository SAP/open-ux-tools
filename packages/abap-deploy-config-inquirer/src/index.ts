import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
