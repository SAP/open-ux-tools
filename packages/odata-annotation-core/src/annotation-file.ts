import type {
    Element,
    Attribute,
    ElementName,
    AttributeName,
    TextNode,
    AnyNode
} from '@sap-ux/odata-annotation-core-types';
import { ELEMENT_TYPE, Edm, TEXT_TYPE } from '@sap-ux/odata-annotation-core-types';

export const elements = (predicate: (child: Element) => boolean | undefined, element: Element): Element[] =>
    (element.content || []).filter(
        (content) => content.type === ELEMENT_TYPE && (!predicate || predicate(content))
    ) as Element[];

export const elementsWithName = (name: ElementName, element: Element): Element[] =>
    elements((content: Element) => content.name === name, element);

export const getElementAttribute = (element: Element, name: AttributeName): Attribute | undefined =>
    element?.attributes?.[name];

export const getElementAttributeValue = (element: Element, name: AttributeName): string => {
    const attributeNode = getElementAttribute(element, name);
    return attributeNode?.value ?? '';
};

/**
 * Determines whether the node is of element type with given name.
 *
 * @param node node to check
 * @param name name to check
 * @returns boolean check result
 */
export function isElementWithName(node: AnyNode | undefined, name: string): node is Element {
    return node?.type === ELEMENT_TYPE && node.name === name;
}

/**
 * Returns true if text doesn't contain any visible charactes (tabs, spaces, line breaks are ignored).
 *
 * @param text
 * @returns boolean check result
 */
function isEmptyText(text: string): boolean {
    return (text || '').replace(/[\s]/g, '').length === 0;
}

/**
 * Get text node content of an element.
 * Elements content is supposed to only contain 'Annotation' tags or single text node.
 *
 * @param element
 * @returns TextNode
 */
export function getSingleTextNode(element: Element): TextNode | null {
    let isInvalid = false;
    let firstTextNode: TextNode | null = null;
    (element.content || []).forEach((node) => {
        if (!isInvalid) {
            if (node.type === ELEMENT_TYPE && node.name !== Edm.Annotation) {
                isInvalid = true; // child element which is not 'Annotation'
            } else if (node.type === TEXT_TYPE && !isEmptyText(node.text)) {
                if (firstTextNode) {
                    isInvalid = true; // second text node
                } else {
                    firstTextNode = node;
                }
            }
        }
    });
    return isInvalid ? null : firstTextNode;
}
