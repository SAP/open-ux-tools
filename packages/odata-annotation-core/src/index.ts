export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    toFullyQualifiedName,
    resolveName,
    getAliasInformation,
    getAllNamespacesAndReferences,
    toAliasQualifiedName
} from './names';

export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath,
    toFullyQualifiedPath
} from './paths';

export { findPathToPosition, FindPathResult, getPositionData } from './search';

export * from './position';

export {
    elements,
    elementsWithName,
    getElementAttribute,
    getElementAttributeValue,
    getSingleTextNode,
    isElementWithName
} from './annotation-file';

export * from './utils';

export * from '@sap-ux/odata-annotation-core-types';
