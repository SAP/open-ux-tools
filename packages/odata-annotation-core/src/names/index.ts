export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    COLLECTION_PREFIX
} from './parse';
export { toFullyQualifiedName, resolveName, toAliasQualifiedName } from './normalization';

export { getAliasInformation, getAllNamespacesAndReferences } from './namespaces';
