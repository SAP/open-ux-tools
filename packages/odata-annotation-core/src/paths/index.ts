export { toFullyQualifiedPath } from './normalization';
export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath
} from './parse';
