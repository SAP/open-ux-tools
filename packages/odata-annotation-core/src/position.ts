import type { Position, Range } from '@sap-ux/odata-annotation-core-types';

/**
 * Checks if position1 is before position2.
 *
 * @param pos1 position1
 * @param pos2 position2
 * @param includeEqual when positions are equal - if true then result is true, otherwise false
 * @returns boolean result
 */
export function isBefore(pos1: Position, pos2: Position, includeEqual = false): boolean {
    if (pos1.line < pos2.line) {
        return true;
    }
    if (pos1.line > pos2.line) {
        return false;
    }
    if (includeEqual) {
        return pos1.character <= pos2.character;
    }

    return pos1.character < pos2.character;
}

/**
 * Checks if position is contained in range.
 *
 * @param range range
 * @param position position
 * @returns boolean result
 */
export function positionContained(range: Range | undefined, position: Position): range is Range {
    return range !== undefined && !isBefore(position, range.start, false) && isBefore(position, range.end, true);
}

/**
 * Checks if position is contained in range (range must be defined).
 *
 * @param range range
 * @param position position
 * @returns boolean result
 */
export function positionContainedStrict(range: Range, position: Position): boolean {
    return !isBefore(position, range.start, false) && isBefore(position, range.end, true);
}
/**
 * Check if the second range is within the first.
 *
 * @param a first range
 * @param b second range
 * @returns booelan result
 */
export function rangeContained(a: Range, b: Range): boolean {
    return isBefore(a.start, b.start, true) && isBefore(b.end, a.end, true);
}

/**
 * Get indent level based on the start position and tab width.
 *
 * @param startPosition
 * @param tabWidth
 * @returns numeric indent level
 */
export function getIndentLevel(startPosition: number, tabWidth: number): number {
    let level: number;
    if (startPosition < 0) {
        level = -1;
    } else {
        level = startPosition / tabWidth;
    }
    return level;
}

/**
 * Indents based on tabs or tab width.
 *
 * @param tabWidth
 * @param useTabs
 * @param level
 * @returns intentation string
 */
export function indent(tabWidth: number, useTabs: boolean, level: number): string {
    if (useTabs) {
        return '\t'.repeat(level);
    } else {
        return ' '.repeat(tabWidth * level);
    }
}
