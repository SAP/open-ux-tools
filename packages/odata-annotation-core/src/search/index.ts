export * from './find-by-position';
