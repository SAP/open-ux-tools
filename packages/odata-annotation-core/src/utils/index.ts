export * from './metadata';
export * from './metadata-service';
