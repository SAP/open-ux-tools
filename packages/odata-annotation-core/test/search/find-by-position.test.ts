import { readFileSync } from 'fs';
import { join } from 'path';

import { Position, Range } from '@sap-ux/odata-annotation-core-types';
import type { AnnotationFile } from '@sap-ux/odata-annotation-core-types';

import type { FindPathResult } from '../../src';
import { findPathToPosition, getPositionData } from '../../src';

function findNodeByPosition(annotationFile: AnnotationFile, position: Position, forCompletion = false): FindPathResult {
    const result = findPathToPosition(annotationFile, position, forCompletion);
    if (result) {
        return getPositionData(annotationFile, result.path);
    }
    return {
        found: false,
        path: '',
        remainingString: '',
        startString: ''
    };
}

const loadTestData = (testName: string) => {
    const folder = join(__dirname, '..', 'data', testName);
    const annoFile = JSON.parse(readFileSync(join(folder, 'annotations.json')).toString()) as AnnotationFile;

    const cursorPosition = JSON.parse(readFileSync(join(folder, 'cursorPosition.json')).toString()) as Position;
    return { annoFile, cursorPosition };
};

const findNodeByPositionTest = (testName: string, forCompletion = false) => {
    const { annoFile, cursorPosition } = loadTestData(testName);
    return findNodeByPosition(annoFile, cursorPosition, forCompletion);
};

const findPathToPositionTest = (testName: string, forCompletion = false) => {
    const { annoFile, cursorPosition } = loadTestData(testName);
    return findPathToPosition(annoFile, cursorPosition, forCompletion);
};

describe('findPathToPosition', () => {
    describe('reference', () => {
        it('reference namespace name', () => {
            const result = findNodeByPositionTest('reference-namespace-name');
            expect(result).toMatchSnapshot();
        });
        it('reference alias', () => {
            const result = findNodeByPositionTest('reference-alias');
            expect(result).toMatchSnapshot();
        });
    });

    describe('target', () => {
        it('target name', () => {
            const result = findNodeByPositionTest('target-name-position');
            expect(result).toMatchSnapshot();
        });

        it('target content', () => {
            const result = findNodeByPositionTest('target-content-position');
            expect(result).toMatchSnapshot();
        });
        it('annotation element name', () => {
            const result = findNodeByPositionTest('annotatations-element-name');
            expect(result).toMatchSnapshot();
        });
    });

    describe('element', () => {
        it('name', () => {
            const result = findNodeByPositionTest('element-name');
            expect(result).toMatchSnapshot();
        });
        it('attribute name', () => {
            const result = findNodeByPositionTest('attribute-name');
            expect(result).toMatchSnapshot();
        });
        it('attribute value', () => {
            const result = findNodeByPositionTest('attribute-value');
            expect(result).toMatchSnapshot();
        });
        it('empty attribute value', () => {
            const result = findNodeByPositionTest('empty-attribute-value');
            expect(result).toMatchSnapshot();
        });
        it('attribute area', () => {
            const result = findNodeByPositionTest('attribute-area');
            expect(result).toMatchSnapshot();
        });
        it('text node value', () => {
            const result = findNodeByPositionTest('text-node-value');
            expect(result).toMatchSnapshot();
        });
        it('element content', () => {
            const result = findNodeByPositionTest('element-content');
            expect(result).toMatchSnapshot();
        });
        it('empty element content', () => {
            const result = findNodeByPositionTest('empty-element-content');
            expect(result).toMatchSnapshot();
        });
        it('element content before annotation', () => {
            const result = findNodeByPositionTest('element-content-before-annotation');
            expect(result).toMatchSnapshot();
        });
        it('element content before annotation for completion', () => {
            const result = findNodeByPositionTest('element-content-before-annotation-for-completion', true);
            expect(result).toMatchSnapshot();
        });
        it('empty element name', () => {
            const result = findNodeByPositionTest('empty-element-name', true);
            expect(result).toMatchSnapshot();
        });
        it('after syntax error', () => {
            const result = findNodeByPositionTest('after-syntax-error', true);
            expect(result).toMatchSnapshot();
        });
        it('unclosed element name', () => {
            const result = findPathToPositionTest('unclosed-element-name', true);
            expect(result).toMatchSnapshot();
        });
        it('content after unclosed element', () => {
            const result = findPathToPositionTest('content-after-unclosed-element', true);
            expect(result).toMatchSnapshot();
        });
        it('element content with no name range', () => {
            const file: AnnotationFile = {
                type: 'annotation-file',
                uri: 'file.cds',
                range: Range.create(0, 0, 2, 2),
                references: [],
                targets: [
                    {
                        type: 'target',
                        name: 'Target1',
                        range: Range.create(0, 0, 2, 2),
                        termsRange: Range.create(1, 1, 2, 2),
                        terms: [
                            {
                                type: 'element',
                                name: '',
                                attributes: {},
                                range: Range.create(1, 1, 2, 2),
                                contentRange: Range.create(1, 1, 2, 2),
                                content: [
                                    {
                                        type: 'text',
                                        text: 'test',
                                        range: Range.create(1, 1, 1, 5)
                                    }
                                ]
                            }
                        ]
                    }
                ]
            };
            const position = Position.create(1, 1);
            const result = findNodeByPosition(file, position, false);
            expect(result).toMatchSnapshot();
        });
    });

    describe('namespace', () => {
        it('schema content', () => {
            const result = findNodeByPositionTest('schema-content');
            expect(result).toMatchSnapshot();
        });
    });
});

describe('getPositionData', () => {
    let annotationFile: AnnotationFile;
    beforeAll(() => {
        ({ annoFile: annotationFile } = loadTestData('get-position-data'));
    });
    describe('basic', () => {
        test('root', () => {
            const positionPointer = '/';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('empty path', () => {
            const positionPointer = '';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('non terminal $', () => {
            const positionPointer = 'targets/$1/name/$20';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toStrictEqual({
                found: false,
                path: "$['targets']['$1']",
                startString: '',
                remainingString: ''
            });
        });
    });
    describe('references', () => {
        test('name', () => {
            const positionPointer = 'references/0/name/$28';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('alias', () => {
            const positionPointer = 'references/0/alias/$8';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
    });
    describe('targets', () => {
        test('name', () => {
            const positionPointer = 'targets/1/name/$20';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('content', () => {
            const positionPointer = 'targets/0/terms';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('annotation element name', () => {
            const positionPointer = 'targets/0/';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
    });
    describe('elements', () => {
        test('name', () => {
            const positionPointer = 'targets/1/terms/0/content/1/content/1/name/$2';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('attribute name', () => {
            const positionPointer = 'targets/1/terms/0/content/1/content/1/content/3/attributes/Path/name/$3';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('attribute value', () => {
            const positionPointer = 'targets/1/terms/0/content/1/content/1/content/3/attributes/Path/value/$1';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('attribute area', () => {
            const positionPointer = 'targets/0/terms/0/attributes/$0';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('text node value', () => {
            const positionPointer = 'targets/2/terms/0/content/1/content/1/content/5/content/1/content/0/text/$4';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('element content', () => {
            const positionPointer = 'targets/1/terms/0/content/1/content/1/content/0/text/$0';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('empty element content', () => {
            const { annoFile } = loadTestData('empty-element-content');
            const positionPointer = 'targets/0/terms/0/content/1/content/$0';
            const result = getPositionData(annoFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
        test('empty element name', () => {
            const { annoFile } = loadTestData('empty-element-name');
            const positionPointer = 'targets/0/terms/0/content/1/name';
            const result = getPositionData(annoFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
    });
    describe('namespace', () => {
        test('schema content', () => {
            const positionPointer = 'namespace/targets';
            const result = getPositionData(annotationFile, positionPointer);
            expect(result).toMatchSnapshot();
        });
    });
});
