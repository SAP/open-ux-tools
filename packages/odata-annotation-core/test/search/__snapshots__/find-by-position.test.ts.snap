// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findPathToPosition element after syntax error 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['content'][1]['attributes']['']['name']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element attribute area 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['attributes']['']['name']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element attribute name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['content'][3]['attributes']['Path']['name']",
  "remainingString": "h",
  "startString": "Pat",
}
`;

exports[`findPathToPosition element attribute value 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['content'][3]['attributes']['Path']['value']",
  "remainingString": "roduct/@UI.CriticalityLabels/0/Label",
  "startString": "P",
}
`;

exports[`findPathToPosition element content after unclosed element 1`] = `
Object {
  "path": "targets/0/terms/0/content/0/text",
  "range": Object {
    "end": Object {
      "character": 4,
      "line": 6,
    },
    "start": Object {
      "character": 36,
      "line": 5,
    },
  },
}
`;

exports[`findPathToPosition element element content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['content'][1]['content'][0]['text']",
  "remainingString": "
                ",
  "startString": "",
}
`;

exports[`findPathToPosition element element content before annotation 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['attributes']['']['name']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element element content before annotation for completion 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element element content with no name range 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][0]['text']",
  "remainingString": "test",
  "startString": "",
}
`;

exports[`findPathToPosition element empty attribute value 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['attributes']['String']['value']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element empty element content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['content'][0]['text']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element empty element name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['name']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition element name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['name']",
  "remainingString": "cord",
  "startString": "Re",
}
`;

exports[`findPathToPosition element text node value 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['content'][1]['content'][1]['content'][1]['content'][0]['text']",
  "remainingString": "egatedPrice",
  "startString": "Aggr",
}
`;

exports[`findPathToPosition element unclosed element name 1`] = `
Object {
  "path": "targets/0/terms/0/content/0/name/$0",
  "range": Object {
    "end": Object {
      "character": 37,
      "line": 5,
    },
    "start": Object {
      "character": 37,
      "line": 5,
    },
  },
}
`;

exports[`findPathToPosition namespace schema content 1`] = `
Object {
  "found": true,
  "path": "$['namespace']['targets']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition reference reference alias 1`] = `
Object {
  "found": true,
  "path": "$['references'][0]['alias']",
  "remainingString": "s",
  "startString": "Analytic",
}
`;

exports[`findPathToPosition reference reference namespace name 1`] = `
Object {
  "found": true,
  "path": "$['references'][0]['name']",
  "remainingString": "cs.v1",
  "startString": "com.sap.vocabularies.Analyti",
}
`;

exports[`findPathToPosition target annotation element name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition target target content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`findPathToPosition target target name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['name']",
  "remainingString": "dents/priority",
  "startString": "IncidentService.Inci",
}
`;

exports[`getPositionData basic empty path 1`] = `
Object {
  "found": true,
  "path": "$",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData basic root 1`] = `
Object {
  "found": true,
  "path": "$['']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData elements attribute area 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['attributes']['$0']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData elements attribute name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['content'][3]['attributes']['Path']['name']",
  "remainingString": "h",
  "startString": "Pat",
}
`;

exports[`getPositionData elements attribute value 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['content'][3]['attributes']['Path']['value']",
  "remainingString": "roduct/@UI.CriticalityLabels/0/Label",
  "startString": "P",
}
`;

exports[`getPositionData elements element content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['content'][0]['text']",
  "remainingString": "
                            ",
  "startString": "",
}
`;

exports[`getPositionData elements empty element content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['content']['$0']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData elements empty element name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms'][0]['content'][1]['name']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData elements name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['terms'][0]['content'][1]['content'][1]['name']",
  "remainingString": "cord",
  "startString": "Re",
}
`;

exports[`getPositionData elements text node value 1`] = `
Object {
  "found": true,
  "path": "$['targets'][2]['terms'][0]['content'][1]['content'][1]['content'][5]['content'][1]['content'][0]['text']",
  "remainingString": "egatedPrice",
  "startString": "Aggr",
}
`;

exports[`getPositionData namespace schema content 1`] = `
Object {
  "found": true,
  "path": "$['namespace']['targets']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData references alias 1`] = `
Object {
  "found": true,
  "path": "$['references'][0]['alias']",
  "remainingString": "s",
  "startString": "Analytic",
}
`;

exports[`getPositionData references name 1`] = `
Object {
  "found": true,
  "path": "$['references'][0]['name']",
  "remainingString": "cs.v1",
  "startString": "com.sap.vocabularies.Analyti",
}
`;

exports[`getPositionData targets annotation element name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData targets content 1`] = `
Object {
  "found": true,
  "path": "$['targets'][0]['terms']",
  "remainingString": "",
  "startString": "",
}
`;

exports[`getPositionData targets name 1`] = `
Object {
  "found": true,
  "path": "$['targets'][1]['name']",
  "remainingString": "dents/priority",
  "startString": "IncidentService.Inci",
}
`;
